### ISSUE TITLE
Simulation of Verilog output doesn't match nMigen simulation

### ISSUE THREAD

**Issue by jeanthom** [nmigen/nmigen#418]
Hi! I was working on a design that used comb statements inside a FSM and noticed that nMigen outputs "always @*" verilog blocks. This causes issues in verilog simulators because the `always` block is only evaluated when a signal of its sensibility list changes.
 
Here's some minimal code that showcases this bug:
```python
from nmigen import *
from nmigen.cli import main

class Top(Elaboratable):
	def __init__(self):
		self.a = Signal()

	def elaborate(self, platform):
		m = Module()

		with m.FSM():
			with m.State("aaaaaaaa"):
				m.d.comb += self.a.eq(1)

		# Workaround for #417 
		b = Signal()
		m.d.sync += b.eq(1)

		return m

if __name__ == "__main__":
	top = Top()
	main(top, ports=[top.a])
```

Its verilog output:
```verilog
/* Generated by Yosys 0.9+2406 (git sha1 9d658a19, gcc 10.1.1 -fPIC -Os) */

(* \nmigen.hierarchy  = "top" *)
(* top =  1  *)
(* generator = "nMigen" *)
module top(b, a);
  (* src = "test.py:6" *)
  output a;
  reg a;
  (* src = "test.py:7" *)
  input b;
  (* src = "test.py:12" *)
  wire fsm_state;
  always @* begin
    a = 1'h0;
    (* src = "test.py:12" *)
    casez (fsm_state)
      /* \nmigen.decoding  = "aaaaaaaa/0" */
      /* src = "test.py:13" */
      1'h0:
          a = 1'h1;
    endcase
  end
  assign fsm_state = 1'h0;
endmodule
```

When the Verilog code is simulated with either Icarus Verilog or Modelsim, `a` is invalid (because its value hasn't been set). When simulated using nMigen built-in simulator, `a` is `1`.

**Comment by jeanthom** [comment]
Adding `proc` to the Yosys script seems to fix the issue.

**Comment by whitequark** [comment]
We should already be handling that case (grep for `$verilog_initial_trigger`...) but we don't because of a bug.

**Comment by jeanthom** [comment]
Is there a ticket in Yosys for the bug you mention?

**Comment by whitequark** [comment]
It is not technically a Yosys bug; Yosys outputs "structural netlists" so it doesn't care about what iverilog does. The bug lies in the way nMigen combines with Yosys. I'll need to think about a good way to fix this.

**Comment by whitequark** [comment]
This is actually a surprisingly subtle issue. I'll raise it at the next Yosys meeting.

**Comment by whitequark** [comment]
@jeanthom I've discussed this in a Yosys meeting today. The conclusion was to adjust the `write_verilog` output to include a workaround for this problem, such that instead of:
```verilog
reg x;
always @* begin
    x <= 1;
end
```
we get something like:
```verilog
reg init;
initial init = 0;
reg x;
always @* begin
    if (init) begin end
    x <= 1;
end
```

**Comment by jeanthom** [comment]
Thank you for taking the time to discuss this issue with the Yosys team. That solution seems good, may I ask you why an `initial` statement wasn't considered (so that the register is set to its correct initial value)?

**Comment by whitequark** [comment]
> That solution seems good, may I ask you why an `initial` statement wasn't considered (so that the register is set to its correct initial value)?

It was; `reg x = 0;` is the same thing as `reg x; initial x = 0;`, and for the simple cases where the initial value does not depend on any other wires/regs in the design, we already do that. But for complex cases like in this bug report, how would you use `initial`?

**Comment by jeanthom** [comment]
I would resolve the wire/regs dependencies, but this might be something to avoid on really large designs. BTW does the `init` approach induces a performance penalty in some simulators?

**Comment by whitequark** [comment]
> I would resolve the wire/regs dependencies

Yeah, so requiring manual changes makes that approach unviable, of course. The goal of a code generator is to make my life simpler, not the other way around.

> BTW does the `init` approach induces a performance penalty in some simulators?

Hard to say without benchmarking. I know it wouldn't affect Verilator.

**Comment by jeanthom** [comment]
The issue was solved upstream.

**Comment by whitequark** [comment]
This is actually not entirely fixed yet--nMigen should detect whether Yosys has the fix, and omit `proc_prune` if it doesn't.

**Comment by whitequark** [comment]
It turns out I've also forgot to fix the Verilog-2001 mode; https://github.com/YosysHQ/yosys/issues/2271#issuecomment-680704258.

**Comment by whitequark** [comment]
> When the Verilog code is simulated with either Icarus Verilog or Modelsim, `a` is invalid (because its value hasn't been set).

Now that I'm looking at this again, I'm confused. How is the behavior of your snippet not a simualtor bug? The continuous assignment `assign fsm_state = 1'h0;` should produce an event at time 0 per IEEE 1364-2005 §11.6.1:

![Screenshot_20200826_085846](https://user-images.githubusercontent.com/54771/91283619-59fcfd80-e77a-11ea-9eeb-1d9cff868cc2.png)

Further, the `always @*` block should be scheduled in response to that event, which is explicitly said in IEEE 1364-2005 §9.7.5:

![Screenshot_20200826_090036](https://user-images.githubusercontent.com/54771/91283809-9af51200-e77a-11ea-9c1d-7a61c5e83249.png)
![Screenshot_20200826_090040](https://user-images.githubusercontent.com/54771/91283818-9df00280-e77a-11ea-98e0-681f2e30d3fc.png)

**Comment by whitequark** [comment]
@jeanthom If I run the following Verilog with Icarus Verilog:

```verilog
/* Generated by Yosys 0.9+2406 (git sha1 9d658a19, gcc 10.1.1 -fPIC -Os) */

(* \nmigen.hierarchy  = "top" *)
(* top =  1  *)
(* generator = "nMigen" *)
module top(b, a);
  (* src = "test.py:6" *)
  output a;
  reg a;
  (* src = "test.py:7" *)
  input b;
  (* src = "test.py:12" *)
  wire fsm_state;
  always @* begin
    a = 1'h0;
    (* src = "test.py:12" *)
    casez (fsm_state)
      /* \nmigen.decoding  = "aaaaaaaa/0" */
      /* src = "test.py:13" */
      1'h0:
          a = 1'h1;
    endcase
  end
  assign fsm_state = 1'h0;

  initial begin
    $dumpfile("t.vcd");
    $dumpvars;
    #100;
    $finish;
  end
endmodule
```

I get this VCD file:

```
$scope module top $end
$var wire 1 ! b $end
$var wire 1 " fsm_state $end
$var reg 1 # a $end
$upscope $end
$enddefinitions $end
#0
$dumpvars
1#
0"
z!
$end
#100
```

That *definitely* sets `a`. Please take another look at this issue.

**Comment by jeanthom** [comment]
I'm also confused because I spent the whole day trying to reproduce the issue with older versions of Yosys to no avail... Probably the minified test code was wrong?

The issue I faced in gram#18 was that the generated Verilog code looked like this:
```verilog
  reg fsm_state;
  always @* begin
    a = 1'h0;
    casez (fsm_state)
      /* \nmigen.decoding  = "aaaaaaaa/0" */
      /* src = "test.py:13" */
      1'h0:
          a = 1'h1;
    endcase
  end
```

`fsm_state` didn't have a default value, thus not triggering an event at t=0. The issue was indeed fixed with your PR on Yosys.
